File: backend/app/utils/telegram.py
import re
from typing import List

import httpx

from app.config import settings

_MD_ESCAPE_RE = re.compile(r"([_*`\[])")


def escape_markdown(value: str) -> str:
    return _MD_ESCAPE_RE.sub(r"\\\1", value)


def format_phone(raw: str) -> str:
    digits = re.sub(r"\D", "", raw)
    if not digits.startswith("998") or len(digits) < 12:
        return raw
    local = digits[3:12]
    return f"+998 {local[0:2]} {local[2:5]} {local[5:7]} {local[7:9]}"


async def send_contact_to_telegram(
    name: str, phone: str, service: str, message: str | None, source: str | None = None
) -> List[str]:
    token = settings.telegram_bot_token
    admin_ids = settings.telegram_admin_ids
    if not token or not admin_ids:
        return []

    safe = {
        "name": escape_markdown(name.strip()),
        "phone": escape_markdown(format_phone(phone)),
        "service": escape_markdown(service.strip()),
        "message": escape_markdown((message or "No message").strip()),
        "source": escape_markdown(source or "fazo.uz"),
    }
    text = (
        "*New Lead*\n"
        f"*Name:* {safe['name']}\n"
        f"*Phone:* {safe['phone']}\n"
        f"*Service:* {safe['service']}\n"
        f"*Source:* {safe['source']}\n"
        f"*Message:* {safe['message']}"
    )

    delivered: List[str] = []
    async with httpx.AsyncClient(timeout=10.0) as client:
        for chat_id in admin_ids:
            try:
                r = await client.post(
                    f"https://api.telegram.org/bot{token}/sendMessage",
                    json={"chat_id": chat_id, "text": text, "parse_mode": "Markdown"},
                )
                if r.status_code == 200 and r.json().get("ok"):
                    delivered.append(chat_id)
            except httpx.HTTPError:
                continue
    return delivered

File: frontend/src/lib/telegram.ts
/**
 * Telegram — OUTBOUND ONLY.
 *
 * This module must NEVER call setWebhook, deleteWebhook, or getUpdates.
 * The same bot token drives the Mudarris CRM via a webhook; a second
 * consumer of updates would break it. Sending messages is safe to share.
 */
import type { Lead } from "@/lib/types/content";

const TELEGRAM_API = "https://api.telegram.org";

function escapeHtml(s: string): string {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

export function formatLeadMessage(lead: Pick<Lead, "name" | "phone" | "service" | "message" | "source" | "project" | "locale">): string {
  const tag = lead.project === "mudarris" ? "🎓 MUDARRIS" : "🚀 FAZO";
  const lines = [
    `<b>${tag} — yangi ariza</b>`,
    "",
    `👤 <b>Ism:</b> ${escapeHtml(lead.name)}`,
    `📞 <b>Telefon:</b> ${escapeHtml(lead.phone)}`,
    `🛠 <b>Xizmat:</b> ${escapeHtml(lead.service || "-")}`,
  ];
  if (lead.message) lines.push(`💬 <b>Xabar:</b> ${escapeHtml(lead.message)}`);
  lines.push(`🌐 <b>Manba:</b> ${escapeHtml(lead.source || "-")} (${lead.locale})`);
  return lines.join("\n");
}

/** Send a message to the configured group. Returns true on success, never throws. */
export async function sendTelegramMessage(text: string): Promise<boolean> {
  const token = process.env.TELEGRAM_BOT_TOKEN;
  const chatId = process.env.TELEGRAM_CHAT_ID;
  if (!token || !chatId) {
    console.warn("[telegram] TELEGRAM_BOT_TOKEN or TELEGRAM_CHAT_ID missing — skipping send");
    return false;
  }
  try {
    const res = await fetch(`${TELEGRAM_API}/bot${token}/sendMessage`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        chat_id: chatId,
        text,
        parse_mode: "HTML",
        disable_web_page_preview: true,
      }),
      cache: "no-store",
    });
    if (!res.ok) console.error("[telegram] sendMessage failed:", res.status, await res.text().catch(() => ""));
    return res.ok;
  } catch (err) {
    console.error("[telegram] sendMessage error:", err);
    return false;
  }
}
